File: ProcessData.py
from __future__ import division
import math
import re

prefix = r'\[(?P<node>[0-9]+) - (?P<thread>[0-9a-f]+)\] \{\w+\}\{legion_prof\}: '


def analyze(data):
    tasks = []
    instances = []
    names = {}
    for line in data:
        match = re.compile(prefix + r'Prof Task Info (?P<tid>[0-9]+) (?P<fid>[0-9]+) (?P<pid>[a-f0-9]+) (?P<create>[0-9]+) (?P<ready>[0-9]+) (?P<start>[0-9]+) (?P<stop>[0-9]+)( (?P<spawn>[0-9]+))?').match(line)
        if match is not None:
            tasks.append({
                "task_id" : long(match.group('tid')),
                "func_id" : int(match.group('fid')),
                "proc_id" : int(match.group('pid'), 16),
                "create" : long(match.group('create'))/1000,
                "ready" : long(match.group('ready'))/1000,
                "start" : long(match.group('start'))/1000,
                "stop" : long(match.group('stop'))/1000,
                "spawn" : None
            })
            if match.group('spawn') is not None:
                tasks[-1]["spawn"] = int(match.group('spawn'))            
        match = re.compile(prefix + r'Prof Task Variant (?P<fid>[0-9]+) (?P<name>[a-zA-Z0-9_]+)').match(line)
        if match is not None:
            names[int(match.group('fid'))] = match.group('name')
        #match = re.compile(prefix + r'Prof Meta Info (?P<opid>[0-9]+) (?P<hlr>[0-9]+) (?P<pid>[a-f0-9]+) (?P<create>[0-9]+) (?P<ready>[0-9]+) (?P<start>[0-9]+) (?P<stop>[0-9]+)').match(line)
        match = re.compile(prefix + r'Prof Inst Info (?P<opid>[0-9]+) (?P<inst>[a-f0-9]+) (?P<mem>[a-f0-9]+) (?P<bytes>[0-9]+) (?P<create>[0-9]+) (?P<destroy>[0-9]+)').match(line)
        if match is not None:
            instances.append({
                "op_id" : long(match.group('opid')),
                "inst_id" : int(match.group('inst'),16),
                "mem_id" : int(match.group('mem'),16),
                "size" : long(match.group('bytes')),
                "create" : match.group('create'),
                "destroy" : match.group('destroy')            
            })
    return tasks, names, instances


def fromFile(fname):
    test = open(fname, "r")
    testdata = test.readlines()
    return analyze(testdata)


if __name__=="__main__":
    # Load the file
    test = open("data/PROFSimple.log", "r")
    testdata = test.readlines()
    result = analyze(testdata)
    print result

File: processdata.js
/**
 * Created by Phil on 4/28/2016.
 */

(function(){
    var prefix = '\\[([0-9]+) - ([0-9a-f]+)] \\{\\w+}\\{legion_prof}: ';

    var processor_kinds = {
        1: 'GPU',
        2: 'CPU',
        3: 'Utility',
        4: 'I/O'
    };

    var procPretty = function(proc){
        var h = Number(proc).toString(16);
        var node = String(parseInt(h.slice(3,7),16));
        proc = String(parseInt(h.slice(7,-1),16));
        return "Node: " + node + " Processor: " + proc;
    };

    var analyze = function(data){
        var tasks = [];
        var taskmap = {};
        var tidnames = {};
        var names = {};
        var tid2vid = {};
        var procs = {};
        var lines = data.split("\n");
        lines.forEach(function(line){
            var full_regex = new RegExp(prefix + 'Prof Task Info ([0-9]+) ([0-9]+) ([a-f0-9]+) ([0-9]+) ([0-9]+) ([0-9]+) ([0-9]+)( ([0-9]+))?');
            var info = full_regex.exec(line);
            if(info != null){
                var task = {
                    "task_id":Number(info[3]),
                    "func_id":Number(info[4]),
                    "proc_id":String(parseInt(info[5],16)),
                    "create":Number(info[6])/1000,
                    "ready":Number(info[7])/1000,
                    "start":Number(info[8])/1000,
                    "stop":Number(info[9])/1000,
                    "spawn":info[10]!= null ? Number(info[10]) : null
                };
                taskmap[task["task_id"]] = task;
                tasks.push(task);
            }
            full_regex = new RegExp(prefix + "Prof Task Kind ([0-9]+) ([a-zA-Z0-9_<>.]+)");
            info = full_regex.exec(line);
            if(info != null){
                tidnames[Number(info[3])] = info[4];
            }
            full_regex = new RegExp(prefix + "Prof Task Variant ([0-9]+) ([0-9]+) ([a-zA-Z0-9_<>.]+)");
            info = full_regex.exec(line);
            if(info != null){
                tidnames[Number(info[3])] = info[4];
            }
            full_regex = new RegExp(prefix + "Prof Proc Desc ([a-f0-9]+) ([0-9]+)");
            info = full_regex.exec(line);
            if(info != null) {
                var kind = Number(info[4]);
                procs[String(parseInt(info[3],16))] = processor_kinds[kind];
            }
        });
        tidnames.forEach(function(n) {
            names[tid2vid[n]] = tidnames[n];
        });
        var proclist = [];
        procs.forEach(function(proc){
            proclist.push({
                'num':Number(proc),
                'id':procPretty(proc),
                'name':procs[proc]
            });
        });
        proclist.sort(function(a,b){
           return a['num'] - b['num'];
        });
        tasks.forEach(function(task){
            task["proc_kind"] = procs[task["proc_id"]];
            task["proc_id"] = procPretty(task["proc_id"])
        });
        return {
            "tasks":tasks,
            "names":names,
            "proclist":proclist
        };
    };

    if (typeof module !== 'undefined' && typeof module.exports !== 'undefined'){
        module.exports = analyze;
    }
    else{
        window.analyze = analyze;
    }
})();
